File: applications/ExternalSolversApplication/custom_external_libraries/FEAST/3.0/example/Hermitian/Fortran/1_dense/helloworld.f90
program helloworld
  implicit none
  !! eigenvalue system 
  integer,parameter :: N=2,LDA=2
  character(len=1),parameter :: UPLO='F' ! 'L' or 'U' also fine
  double precision,dimension(N*N) :: A=(/2.0d0,-1.0d0,-1.0d0,2.0d0/)
  double precision :: Emin=-5.0d0, Emax=5.0d0
  integer :: M0=2 ! (Initial) subspace dimension
  !! input parameters for FEAST
  integer,dimension(64) :: feastparam
  !! output variables for FEAST
  double precision,dimension(:),allocatable :: E, res
  double precision,dimension(:,:),allocatable :: X
  double precision :: epsout
  integer :: i,loop,info,M

!!! Allocate memory for eigenvalues.eigenvectors/residual
  allocate(E(M0))
  allocate(res(M0))
  allocate(X(N,M0))

!!!!!!!!!! FEAST
  call feastinit(feastparam)
  feastparam(1)=1 !! change from default value
  call dfeast_syev(UPLO,N,A,LDA,feastparam,epsout,loop,Emin,Emax,M0,E,X,M,res,info)

!!!!!!!!!! REPORT
  print *,'FEAST OUTPUT INFO',info
  if (info==0) then
     print *,'*************************************************'
     print *,'************** REPORT ***************************'
     print *,'*************************************************'
     print *,'# Search interval [Emin,Emax]',Emin,Emax
     print *,'# mode found/subspace',M,M0
     print *,'# iterations',loop
     print *,'TRACE',sum(E(1:M))
     print *,'Relative error on the Trace',epsout
     print *,'Eigenvalues/Residuals'
     do i=1,M
        print *,i,E(i),res(i)
     enddo
     print *,'Eigenvectors'
     do i=1,M
        print *,i,"(",X(1,i),X(2,i),")"
     enddo
  endif

end program helloworld

File: applications/ExternalSolversApplication/custom_external_libraries/FEAST/3.0/example/Hermitian/C/1_dense/helloworld.c
#include <stdio.h> 
#include <stdlib.h> 

#include "feast.h"
#include "feast_dense.h"
int main() { 
  /*  eigenvalue system  */
  int   N=2,LDA=2;
  char  UPLO='F'; // 'L' and 'U' also fine
  double A[4]={2.0,-1.0,-1.0,2.0}; 
  double Emin=-5.0, Emax=5.0;
  int M0=2; //size initial subspace
  /* input parameters for FEAST */
  int feastparam[64];
  /* output variables for FEAST */
  double *E, *res, *X; 
  double  epsout,trace; 
  int i,loop,info,M;

  /* Allocate memory for eigenvalues.eigenvectors/residual */
  E=calloc(M0,sizeof(double));  //eigenvalues
  res=calloc(M0,sizeof(double));//eigenvectors
  X=calloc(N*M0,sizeof(double));//residual 

  /* !!!!!!!!!! FEAST !!!!!!!!!*/
  feastinit(feastparam); 
  feastparam[0]=1;  /*change from default value */
  dfeast_syev(&UPLO,&N,A,&LDA,feastparam,&epsout,&loop,&Emin,&Emax,&M0,E,X,&M,res,&info);

  /*!!!!!!!!!! REPORT !!!!!!!!!*/
  printf("FEAST OUTPUT INFO %d\n",info);
  if (info==0) {
    printf("*************************************************\n");
    printf("************** REPORT ***************************\n");
    printf("*************************************************\n");
    printf("# Search interval [Emin,Emax] %.15e %.15e\n",Emin,Emax);
    printf("# mode found/subspace %d %d \n",M,M0);
    printf("# iterations %d \n",loop);
    trace=0.0;
    for (i=0;i<=M-1;i=i+1){
      trace=trace+*(E+i);
    }	  
    printf("TRACE %.15e\n", trace);
    printf("Relative error on the Trace %.15e\n",epsout );
    printf("Eigenvalues/Residuals\n");
    for (i=0;i<=M-1;i=i+1){
      printf("   %d %.15e %.15e\n",i,*(E+i),*(res+i));
    }
    printf("Eigenvectors\n");
    for (i=0;i<=M-1;i=i+1){
      printf("   %d (%.15e, %.15e)\n",i,*(X+i*M),*(X+1+i*M));
    }
  }
free(E);
free(res);
free(X);
  return 0;
}




